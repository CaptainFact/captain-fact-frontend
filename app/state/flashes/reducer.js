import { List, Record } from 'immutable'
import { handleActions, createAction, combineActions } from 'redux-actions'
import { buildFlash } from './record'
import { getErrorInfo } from '../../lib/errors'
import { NO_INTERNET_ERROR } from '../../constants'


// Actions
export const addFlash = createAction('FLASHES/ADD', options => buildFlash(options))
export const removeFlash = createAction('FLASHES/REMOVE')
export const pause = createAction('FLASHES/PAUSE', () => true)
export const unPause = createAction('FLASHES/UNPAUSE', () => false)
export const update = createAction('FLASHES/UPDATE')

// Actions helpers (use them like regular actions)
export const flashError = options => addFlash({
  flashType: 'danger',
  iconName: 'exclamation-circle',
  ...options
})
export const flashErrorMsg = message => flashError({message})
export const flashErrorUnauthenticated = () => flashError({
  message: 'errors:server.unauthenticated',
  infoText: 'main:menu.loginSignup',
  infoUrl: '/login'
})

export const flashSuccessMsg = (message, params = {}) => addFlash({
  flashType: 'success',
  iconName: 'check-circle',
  message,
  ...params
})
export function errorToFlash(msg) {
  const errorInfo = getErrorInfo(msg)
  let action = null
  if (!errorInfo)
    action = flashError({message: msg, isError: true})
  else
    action = flashError({
      message: msg,
      infoUrl: errorInfo.url,
      infoText: errorInfo.i18nKey,
      isError: true
    })
  action.error = true
  return action
}
// Same as errorToFlash but doesn't show anything if payload is
// not a string (useful for forms)
export function errorMsgToFlash(msg) {
  if (typeof (msg) === 'string')
    return errorToFlash(msg)
  return () => () => null
}

// Reducer
const INITIAL_STATE = Record({
  flashes: new List(),
  isPaused: false
})

const FlashesReducer = handleActions({
  [addFlash]: (state, {payload}) => {
    // Only display one error for connections problems (instead of one per request)
    if (payload.message === NO_INTERNET_ERROR
      && state.flashes.find(f => f.message === NO_INTERNET_ERROR))
      return state
    return state.update('flashes', l => l.push(payload))
  },
  [removeFlash]: (state, {payload: {id}}) => {
    const flashIdx = state.flashes.findIndex(msg => msg.id === id)
    if (flashIdx !== -1)
      return state.update('flashes', l => l.delete(flashIdx))
    return state
  },
  [combineActions(pause, unPause)]: (state, {payload}) => state.set('isPaused', payload),
  [update]: (state, {payload}) => {
    if (!state.isPaused)
      return state.update('flashes', flashes => flashes
        .map(f => f.set('timeLeft', f.timeLeft - payload))
        .filter(msg => msg.timeLeft > 0)
      )
    return state
  }
}, INITIAL_STATE())

export default FlashesReducer
