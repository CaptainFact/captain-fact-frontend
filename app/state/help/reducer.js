import { Record } from "immutable"
import { createAction, handleActions } from 'redux-actions'


export const setLoading = createAction('HELP/SET_LOADING')
export const reset = createAction('HELP/RESET')
export const setContent = createAction('HELP/SET_HELP_PAGE')

const INITIAL_STATE = new Record({
  markdownContent: "",
  isLoading: false,
  error: null
})

const HelpReducer = handleActions({
  [setContent]: {
    next: (state, {payload}) => state.merge({
      markdownContent: payload,
      isLoading: false
    }),
    throw: (state, {payload}) => state.merge({error: payload, isLoading: false})
  },
  [setLoading]: (state, {payload}) => state.set('isLoading', payload),
  [reset]: () => INITIAL_STATE()
}, INITIAL_STATE())

export default HelpReducer
