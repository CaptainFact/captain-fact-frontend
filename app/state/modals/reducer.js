import { Stack } from 'immutable'
import { createAction, handleActions } from 'redux-actions'


export const addModal = createAction('ADD_MODAL')
export const popModal = createAction('POP_MODAL')

const ModalsReducer = handleActions({
  [addModal]: (state, {payload: {Modal, props}}) => state.push({Modal, props}),
  [popModal]: state => state.pop()
}, new Stack())

export default ModalsReducer
