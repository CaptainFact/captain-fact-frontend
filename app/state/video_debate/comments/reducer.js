import { Record, Map, List, Set } from 'immutable'
import { createAction, handleActions } from 'redux-actions'
import isAfter from 'date-fns/is_after'

import Comment from './record'
import parseDateTime from '../../../lib/parse_datetime'
import { resetVideoDebate } from '../actions'


export const addFlag = createAction('COMMENTS/ADD_FLAG')

export const addMyVote = createAction('COMMENTS/ADD_MY_VOTE')
export const setVoting = createAction('COMMENTS/SET_VOTING')
export const endVoting = createAction('COMMENTS/END_VOTING')

export const fetchAll = createAction('COMMENTS/FETCH_ALL')
export const update = createAction('COMMENTS/UPDATE')
export const updateScores = createAction('COMMENTS/UPDATE_SCORES')
export const setLoading = createAction('COMMENTS/SET_LOADING')
export const remove = createAction('COMMENTS/DELETE')
export const add = createAction('COMMENTS/ADD')

const INITIAL_STATE = new Record({
  isLoading: false,
  errors: null,
  comments: new Map(),
  replies: new Map(),
  voted: new Map(),
  voting: new Set(),
  myFlags: new Set()
})

const CommentsReducer = handleActions({
  // Flags
  [addFlag]: (state, {payload}) => state.update('myFlags', f => f.add(payload)),
  // Votes
  [addMyVote]: (state, {payload: {comment, value}}) => {
    const prevVote = state.voted.get(comment.id, 0)
    const voteChange = value === 0 ?
      -prevVote : Math.abs(prevVote - value) * (value >= 0 ? 1 : -1)
    const commentPath = getCommentPath(comment)
    const commentIdx = getCommentIdx(state, commentPath, comment.id)
    if (commentIdx !== -1)
      state = state
        .updateIn([...commentPath, commentIdx, 'score'], s => s + voteChange)
        .updateIn(commentPath, comments => comments.sort(commentsComparator))

    return state
      .setIn(['voted', comment.id], value)
      .update('voting', s => s.delete(comment.id))
  },
  [setVoting]: (state, {payload}) => state.update('voting', s => s.add(payload)),
  [endVoting]: (state, {payload}) => state.update('voting', s => s.remove(payload)),
  // Comments
  [setLoading]: (state, {payload}) => state.set('isLoading', payload),
  [fetchAll]: {
    next: (state, {payload: {comments, my_votes, my_flags}}) => {
      const preparedComments =
        new List(comments.map(prepareComment)
          .sort(commentsComparator))
          .groupBy(c => c.reply_to_id)

      const replies = preparedComments.delete(null)
      const notReplies = preparedComments.get(null) || new List()

      return state.merge({
        comments: notReplies.groupBy(c => c.statement_id),
        errors: null,
        isLoading: false,
        replies,
        voted: !my_votes ? new Map() : new Map().withMutations(map =>
          my_votes.forEach(vote => map.set(vote.comment_id, vote.value))
        ),
        myFlags: new Set(my_flags)
      })
    },
    throw: (state, {payload}) => state.merge({errors: payload, isLoading: false})
  },
  [add]: (state, {payload}) => {
    const comment = prepareComment(payload)
    const commentPath = getCommentPath(comment)
    const commentsList = state.getIn(commentPath, new List())
    const insertIdx = commentsList.findIndex(c => commentsComparator(comment, c) < 0)
    const newCommentsList = commentsList.insert(
      insertIdx !== -1 ? insertIdx : commentsList.size,
      comment
    )
    return state.setIn(commentPath, newCommentsList)
  },
  [update]: (state, {payload}) => {
    const commentFullPath = getCommentFullPath(state, payload)
    if (!commentFullPath)
      return state
    const comment = payload.__partial ? payload : prepareComment(payload)
    return state.updateIn(commentFullPath, c => c.merge(comment))
  },
  [remove]: (state, {payload}) => {
    const commentFullPath = getCommentFullPath(state, payload)
    if (!commentFullPath)
      return state
    return state.deleteIn(commentFullPath)
  },
  [updateScores]: (state, {payload}) => {
    const {comments, replies} =
    new List(payload).groupBy(c => (c.reply_to_id ? 'replies' : 'comments')).toObject()

    // Update comments
    if (comments) {
      const groupedComments = comments.groupBy(c => c.statement_id).entries()
      for (const [statementId, newComments] of groupedComments)
        state = state.updateIn(['comments', statementId], oldList =>
          mergeCommentsList(oldList, newComments)
        )
    }
    // Update replies
    if (replies) {
      const groupedReplies = replies.groupBy(r => r.reply_to_id).entries()
      for (const [commentId, newComments] of groupedReplies)
        state = state.updateIn(['replies', commentId], oldList =>
          mergeCommentsList(oldList, newComments)
        )
    }
    return state
  },
  [resetVideoDebate]: () => INITIAL_STATE()
}, INITIAL_STATE())

// Sort comments by score. More recents come firsts if same score
function commentsComparator(c1, c2) {
  if (c1.score === c2.score)
    return isAfter(c1.inserted_at, c2.inserted_at) ? -1 : 1
  return c1.score > c2.score ? -1 : 1
}

function prepareComment(comment) {
  comment.inserted_at = parseDateTime(comment.inserted_at)
  comment.score = comment.score || 0
  return new Comment(comment)
}

function getCommentFullPath(state, comment) {
  const commentPath = getCommentPath(comment)
  const commentIdx = getCommentIdx(state, commentPath, comment.id)
  if (commentIdx === -1)
    return null
  commentPath.push(commentIdx)
  return commentPath
}

function getCommentPath(comment) {
  return comment.reply_to_id ?
    ['replies', comment.reply_to_id] : ['comments', comment.statement_id]
}

function getCommentIdx(state, path, id) {
  return state.getIn(path, new List()).findIndex(c => c.id === id)
}

function mergeCommentsList(_oldList, newList, addNewComments = false) {
  return _oldList
    .withMutations(oldList => {
      for (const comment of newList) {
        const commentIdx = oldList.findIndex(c => c.id === comment.id)
        if (commentIdx !== -1)
          oldList.update(commentIdx, c => c.merge(comment))
        else if (addNewComments)
          oldList.push(comment)
      }
      return oldList
    })
    .sort(commentsComparator)
}

export default CommentsReducer
