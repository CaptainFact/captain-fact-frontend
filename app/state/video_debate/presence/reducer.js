import { handleActions, createAction } from 'redux-actions'
import { Record } from 'immutable'


export const setPresence = createAction('PRESENCE/SET')
export const presenceDiff = createAction('PRESENCE/DIFF')


const INITIAL_STATE = new Record({
  viewers: new Record({count: 0})(),
  users: new Record({count: 0})(),
})
const PresenceReducer = handleActions({
  [setPresence]: (state, {payload}) => state.merge(payload),
  [presenceDiff]: (state, {payload: {leaves, joins}}) => {
    return state.withMutations(record => record
      .updateIn(['viewers', 'count'], x =>
        (x + joins.viewers.count) - leaves.viewers.count
      )
      .updateIn(['users', 'count'], x =>
        (x + joins.users.count) - leaves.users.count
      )
    )
  }
}, INITIAL_STATE())

export default PresenceReducer
