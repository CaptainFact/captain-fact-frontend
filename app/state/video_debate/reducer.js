import { combineReducers } from "redux"
import PresenceReducer from './presence/reducer'

import VideoReducer from './video/reducer'
import CommentsReducer from './comments/reducer'
import StatementsReducer  from './statements/reducer'


const VideoDebateReducer = combineReducers({
  presence: PresenceReducer,
  video: VideoReducer,
  statements: StatementsReducer,
  comments: CommentsReducer
})
export default VideoDebateReducer
