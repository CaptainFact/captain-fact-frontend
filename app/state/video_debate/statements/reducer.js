import { Record, List } from 'immutable'
import { combineActions, createAction, handleActions } from 'redux-actions'
import { change, destroy } from 'redux-form'

import Statement from './record'
import { resetVideoDebate } from '../actions'


export const fetchStatements = createAction('STATEMENTS/FETCH')
export const setLoading = createAction('STATEMENTS/SET_LOADING')
export const setSubmitting = createAction('STATEMENTS/SET_SUBMITTING')
export const add = createAction('STATEMENTS/ADD')
export const update = createAction('STATEMENTS/UPDATE')
export const updateAll = createAction('STATEMENTS/UPDATE_ALL')
export const remove = createAction('STATEMENTS/REMOVE')
export const setScrollTo = createAction('STATEMENTS/SET_SCROLL_TO')
export const incrementFormCount = createAction('STATEMENTS/INCREMENT_FORM_COUNT', () => 1)
export const decrementFormCount = createAction('STATEMENTS/DECREMENT_FORM_COUNT', () => -1)

// Statement form actions
export const STATEMENT_FORM_NAME = 'StatementForm'
export const changeStatementFormSpeaker = ({id}) => change(STATEMENT_FORM_NAME, 'speaker_id', id)
export const closeStatementForm = () => destroy(STATEMENT_FORM_NAME)


const INITIAL_STATE = new Record({
  isLoading: false,
  isSubmitting: false,
  errors: null,
  data: new List(),
  scrollTo: null,
  formsCount: 0
})

const StatementsReducer = handleActions({
  [fetchStatements]: {
    next: (state, {payload}) => state.merge({
      data: new List(payload.map(s => new Statement(s))).sortBy(st => st.time),
      isLoading: false
    }),
    throw: (state, {payload}) => state.merge({isLoading: false, errors: payload})
  },
  [setLoading]: (state, isLoading) =>
    state.set('isLoading', isLoading),
  [setSubmitting]: (state, {payload}) => state.set('isSubmitting', payload),
  [add]: (state, {payload}) => {
    const statementIdx = getInsertPosition(state.data, payload)
    const statement = new Statement(payload)
    return state.update('data', statements => statements.insert(statementIdx, statement))
  },
  [update]: (state, {payload}) => {
    const statementIdx = state.data.findIndex(s => s.id === payload.id)
    if (statementIdx !== -1)
      return state.update('data', data => {
        data = data.delete(statementIdx)
        const newStatementIdx = getInsertPosition(data, payload)
        return data.insert(newStatementIdx, new Statement(payload))
      })
    return state
  },
  [updateAll]: (state, {payload}) => {
    return state.update('data', data => data.withMutations(statements => {
      // Update all statements
      for (const newStatement of payload) {
        const statementIdx = statements.findIndex(s => s.id === newStatement.id)
        if (statementIdx !== -1)
          statements.update(statementIdx, oldStatement =>
            oldStatement.merge(newStatement)
          )
      }
      // Re-sort them
      return statements.sortBy(st => st.time)
    }))
  },
  [remove]: (state, {payload: {id}}) => {
    const statementIdx = state.data.findIndex(s => s.id === id)
    if (statementIdx !== -1)
      return state.update('data', data => data.delete(statementIdx))
    return state
  },
  [setScrollTo]: (state, {payload}) => state.set('scrollTo', payload),
  [combineActions(incrementFormCount, decrementFormCount)]: (state, {payload}) =>
    state.set('formsCount', state.formsCount + payload),
  [resetVideoDebate]: () => INITIAL_STATE()
}, INITIAL_STATE())

function getInsertPosition(data, newStatement) {
  const position = data.findIndex(st => newStatement.time < st.time)
  return position !== -1 ? position : data.size
}

export default StatementsReducer
