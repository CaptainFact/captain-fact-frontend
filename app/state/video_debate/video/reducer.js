import { handleActions, createAction } from 'redux-actions'
import { Record, List } from 'immutable'
import uuidv1 from 'uuid/v1'

import Video from '../../videos/record'
import Speaker from '../../speakers/record'
import { resetVideoDebate } from '../actions'


export const fetchAll = createAction('VIDEO/FETCH_ALL')
export const setLoading = createAction('VIDEO/SET_LOADING')
export const addSpeaker = createAction('VIDEO/ADD_SPEAKER')
export const removeSpeaker = createAction('VIDEO/REMOVE_SPEAKER')
export const updateSpeaker = createAction('VIDEO/UPDATE_SPEAKER')

export const setPosition = createAction('PLAYBACK/SET_POSITION')
export const forcePosition = createAction('PLAYBACK/FORCE_POSITION')


const FORCED_POSITION_RECORD = new Record({requestId: null, time: 0})

const INITIAL_STATE = new Record({
  data: new Video(),
  errors: null,
  isLoading: false,
  playback: new Record({
    position: null,
    forcedPosition: FORCED_POSITION_RECORD()
  })()
})
const VideoReducer = handleActions({
  [fetchAll]: {
    next: (state, {payload: {speakers, ...data}}) => {
      speakers = sortSpeakers(new List(speakers.map(s => new Speaker(s))))
      return state.mergeDeep({
        isLoading: false,
        errors: null,
        data: new Video(data).set('speakers', speakers)
      })
    },
    throw: (state, {payload}) => state.merge({errors: payload, isLoading: false})
  },
  [setLoading]: (state, {payload}) => state.set('isLoading', payload),
  [addSpeaker]: (state, {payload}) => state.updateIn(['data', 'speakers'], s => sortSpeakers(s.push(new Speaker(payload)))),
  [removeSpeaker]: (state, {payload: {id}}) => {
    const speakerIdx = state.data.speakers.findIndex(s => s.id === id)
    if (speakerIdx !== -1)
      return state.deleteIn(['data', 'speakers', speakerIdx])
    return state
  },
  [updateSpeaker]: (state, {payload}) => {
    const speakerIdx = state.data.speakers.findIndex(s => s.id === payload.id)
    if (speakerIdx !== -1)
      return state
        .mergeIn(['data', 'speakers', speakerIdx], payload)
        .updateIn(['data', 'speakers'], speakers => sortSpeakers(speakers))
    return state
  },
  [resetVideoDebate]: () => INITIAL_STATE(),
  [setPosition]: (state, {payload}) => state.setIn(['playback', 'position'], Math.trunc(payload)),
  [forcePosition]: (state, {payload}) => state.update('playback', p => p.mergeDeep({
    position: payload,
    forcedPosition: {time: payload, requestId: uuidv1()}
  }))
}, INITIAL_STATE())
export default VideoReducer

function sortSpeakers(speakers) {
  return speakers.sortBy(s => s.title + s.full_name)
}
