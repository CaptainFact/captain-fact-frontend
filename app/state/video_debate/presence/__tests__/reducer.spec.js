import reducer, { setPresence, presenceDiff } from '../reducer'


const INITIAL_STATE = reducer(undefined, {})

const mockPresenceEvent = ({
  userJoin = 0,
  viewerJoin = 0,
  userLeave = 0,
  viewerLeave = 0
}) => ({
  joins: {
    viewers: {count: viewerJoin},
    users: {count: userJoin},
  },
  leaves: {
    viewers: {count: viewerLeave},
    users: {count: userLeave},
  }
})

test('start at 0', () => snapshot(INITIAL_STATE))

test('set presence', () => {
  snapshotReducer(reducer, INITIAL_STATE,
    setPresence({viewers: {count: 42}, users: {count: 5}})
  )
})

test('presenceDiff', () => {
  snapshotReducer(reducer, INITIAL_STATE,
    presenceDiff(mockPresenceEvent({userJoin: 2, viewerJoin: 5})),
    presenceDiff(mockPresenceEvent({userLeave: 2, viewerJoin: 5})),
    presenceDiff(mockPresenceEvent({viewerLeave: 10})),
  )
})
