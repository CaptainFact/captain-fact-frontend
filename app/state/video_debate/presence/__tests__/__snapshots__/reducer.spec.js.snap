// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`presenceDiff 1`] = `
Immutable.Record {
  "viewers": Immutable.Record {
    "count": 0,
  },
  "users": Immutable.Record {
    "count": 0,
  },
}
`;

exports[`presenceDiff 2`] = `
Immutable.Record {
  "viewers": Immutable.Record {
    "count": 5,
  },
  "users": Immutable.Record {
    "count": 2,
  },
}
`;

exports[`presenceDiff 3`] = `
Immutable.Record {
  "viewers": Immutable.Record {
    "count": 10,
  },
  "users": Immutable.Record {
    "count": 0,
  },
}
`;

exports[`presenceDiff 4`] = `
Immutable.Record {
  "viewers": Immutable.Record {
    "count": 0,
  },
  "users": Immutable.Record {
    "count": 0,
  },
}
`;

exports[`set presence 1`] = `
Immutable.Record {
  "viewers": Immutable.Record {
    "count": 0,
  },
  "users": Immutable.Record {
    "count": 0,
  },
}
`;

exports[`set presence 2`] = `
Immutable.Record {
  "viewers": {
    "count": 42,
  },
  "users": {
    "count": 5,
  },
}
`;

exports[`start at 0 1`] = `
Immutable.Record {
  "viewers": Immutable.Record {
    "count": 0,
  },
  "users": Immutable.Record {
    "count": 0,
  },
}
`;
